### PR TITLE
Fixes #14199 and other Ctrl-C issues in coqtop

### LINKED ISSUE #14199: Anomaly when quitting coqtop with ^C^D
`Error: Anomaly "Uncaught exception Not_found."` in gramlib
Seems to date from https://github.com/coq/coq/pull/14075

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix

Fixes / closes #14199 as well as other issues with Ctrl-C in coqtop

At the current time, the PR produces the following:
```
Coq < some-text-then-interrupt^CError:
User interrupt.
```
which is not very nicely presented. We would at least like to have `Error:` on the next line, or no `Error:` at all. I don't know well how to do that. See discussion at #14199. Cc @ejgallego.

### CODE REVIEW

File: toplevel/coqloop.ml [hunk 1]
```diff
@@ -266,7 +266,7 @@ let read_sentence ~state input =
   try Stm.parse_sentence ~doc:state.doc state.sid ~entry:G_toplevel.vernac_toplevel input
   with reraise ->
     let reraise = Exninfo.capture reraise in
-    discard_to_dot ();
+    if fst reraise <> Sys.Break then discard_to_dot ();
```
Comment (SkySkimmer) [inline toplevel/coqloop.ml:269]: could use a comment imo

File: gramlib/grammar.ml [hunk 2]
```diff
@@ -1593,6 +1593,7 @@ module Parsable = struct
     | Stream.Error _ as exc ->
       let loc = get_loc () in
       Loc.raise ~loc exc
+    | Sys.Break -> raise Sys.Break (* No location on Sys.Break *)
```
Comment (ppedrot) [inline gramlib/grammar.ml:1596]: More generally, shouldn't we protect against all asynchronous exceptions?

Comment (herbelin) [reply gramlib/grammar.ml:1596]: Do you mean `Out_of_memory` and `Stack_overflow`? Maybe it is ok to leave them with a location attached to them, so as to explain what token generated them?

File: toplevel/coqloop.ml (hunk 1 above)
Comment (herbelin) [reply toplevel/coqloop.ml:278]: I added a comment. Please tell if ok for you.

**Comment by herbelin** [comment]
To improve the reporting when Ctrl-C is hit, I added a newline before `Error:` is printed rather than between `Error:` and `User interrupt.`. (I'm still unsure Ctrl-C should be thought as an error rather than, say, as a special case of result, but at least this is now aesthetically acceptable.)

The current situation is also that the interruption of a computation is located but the interruption of a parsing is not. Don't know if this makes sense, but it is at least acceptable.

File: gramlib/grammar.ml (hunk 2 above)
Comment (ejgallego) [reply gramlib/grammar.ml:1596]: Also in this code we should use `iraise` I think, the original use of Loc.raise was due to camlp5 not having access to Coq machinery IIRC.

Comment (herbelin) [reply gramlib/grammar.ml:1596]: You mean `Exninfo.iraise (Exninfo.capture Sys.Break)`??

**Comment by ppedrot** [comment]
This seems OK as a workaround, I am merging soon except if somebody complains.

**Review by ppedrot — APPROVED** [review]

**Pull request merged**